LFS-34: UI for granting user accounts

veronikaslc opened the pull request:
Used material-table Datatable for table generation

Commit:
LFS-34/35: User Management Page
Used material-table Datatable for compact table generation and nice UI

OrdiNeu commented on modules/principals/src/main/frontend/src/Userboard/Groups/addusertogroupdialogue.jsx:
```diff
@@ -0,0 +1,187 @@
+
+import MaterialTable from 'material-table';
+
+class AddUserToGroupDialogue extends React.Component {
```
In general we've moved to using functional components rather than class components.

marta- replied:
+1 for functional components. I think @veronikaslc inherited the class based components from the initial version. Would be nice to try to switch.

marta- replied:
Per in-person discussion, the switch to functional components is out of the scope of this sprint.

OrdiNeu commented on modules/principals/src/main/frontend/src/Userboard/Groups/addusertogroupdialogue.jsx:
```diff
@@ -0,0 +1,187 @@
+            .then((data) => {
+                var names = [];
+                for (var username in data) {
+                    !username.startsWith("sling-") && names.push(this.addName(username));
```
The entire loop could be replaced with the more idiomatic:
```suggestion
                    names = data.filter((username) => !username.startsWith("sling-")).map((username) => this.addName(username));
```

OrdiNeu commented on modules/principals/src/main/frontend/src/Userboard/Groups/addusertogroupdialogue.jsx:
```diff
@@ -0,0 +1,187 @@
+
+    // TODO - find more efficient way to add and remove users from list
+    handleSelectRowClick(rows) {
+        let chosens = [];
```
Most of this function could be replaced with the simpler
```suggestion
        let chosens = rows.map((row) => row.name);
```

OrdiNeu commented on modules/principals/src/main/frontend/src/Userboard/Groups/addusertogroupdialogue.jsx:
```diff
@@ -0,0 +1,187 @@
+              }
+
+              let filtered = [];
+              filtered = this.state.allUserNames.filter( function( el ) {
```
We should be consistent in our usage of arrow functions.
```suggestion
              filtered = this.state.allUserNames.filter( (el) => {
```

OrdiNeu commented on modules/principals/src/main/frontend/src/Userboard/Groups/addusertogroupdialogue.jsx:
```diff
@@ -0,0 +1,187 @@
+                </DialogContent>
+                <DialogActions>
+                    <Button variant="contained" size="small" color="primary" onClick={() => this.handleAddUsers()} disabled={this.state.selectedUsers.length == 0}>Add</Button>
+                    <Button variant="contained" size="small" onClick={() => this.handleExit()}>Close</Button>
```
Arrow function shouldn't be necessary here
```suggestion
                    <Button variant="contained" size="small" onClick={this.handleExit}>Close</Button>
```

OrdiNeu commented on modules/principals/src/main/frontend/src/Userboard/Groups/addusertogroupdialogue.jsx:
```diff
@@ -0,0 +1,187 @@
+                    </Grid>
+                </DialogContent>
+                <DialogActions>
+                    <Button variant="contained" size="small" color="primary" onClick={() => this.handleAddUsers()} disabled={this.state.selectedUsers.length == 0}>Add</Button>
```
```suggestion
                    <Button variant="contained" size="small" color="primary" onClick={this.handleAddUsers} disabled={this.state.selectedUsers.length == 0}>Add</Button>
```

OrdiNeu commented on modules/principals/src/main/frontend/src/Userboard/Groups/creategroupdialogue.jsx:
```diff
@@ -0,0 +1,78 @@
+        return (
+            <Dialog
+                open={this.props.isOpen}
+                onClose={() => this.props.handleClose()}
```
```suggestion
                onClose={this.props.handleClose}
```

OrdiNeu commented on modules/principals/src/main/frontend/src/Userboard/Groups/creategroupdialogue.jsx:
```diff
@@ -0,0 +1,78 @@
+                </DialogContent>
+                <DialogActions>
+                    <Button color="primary" variant="contained" size="small" onClick={(event) => { event.preventDefault(); this.handleCreateGroup(); }}>Create Group</Button>
+                    <Button variant="contained" size="small" onClick={() => this.props.handleClose()}>Close</Button>
```
```suggestion
                    <Button variant="contained" size="small" onClick={this.props.handleClose}>Close</Button>
```

OrdiNeu commented on modules/principals/src/main/frontend/src/Userboard/Groups/deletegroupdialogue.jsx:
```diff
@@ -0,0 +1,68 @@
+        return (
+            <Dialog
+                open={this.props.isOpen}
+                onClose={() => this.props.handleClose()}
```
```suggestion
                onClose={this.props.handleClose}
```

OrdiNeu commented on modules/principals/src/main/frontend/src/Userboard/Groups/groupsmanager.jsx:
```diff
@@ -0,0 +1,209 @@
+              { actionsColumnIndex: -1 },
+              { headerStyle: { backgroundColor: '#fafbfc'} },
+              { emptyRowsWhenPaging: false }
+            }
```
This appears to be buggy -- you want to pass an object to options, instead of four objects seperated by commas. As it is, only the last option (`emptyRowsWhenPaging: false`) is being used.

OrdiNeu commented on modules/principals/src/main/frontend/src/Userboard/Groups/groupsmanager.jsx:
```diff
@@ -0,0 +1,209 @@
+                icon: 'edit',
+                tooltip: 'Edit Group',
+                onClick: (event, rowData) => this.handleGroupRowClick(rowData.tableData.id, rowData.name)
+              },*/
```
Debug code, remove

OrdiNeu commented on modules/principals/src/main/frontend/src/Userboard/Groups/removeuserfromgroup.jsx:
```diff
@@ -0,0 +1,145 @@
+
+  addName(name) {
+    return { name }
+  }
```
What's this function used for?

OrdiNeu commented on modules/principals/src/main/frontend/src/Userboard/Groups/removeuserfromgroup.jsx:
```diff
@@ -0,0 +1,145 @@
+      });
+  }
+
+  handleSelectRowClick(rows) {
```
There appears to be a lot of code duplication between this and addUserToGroupDialog.

marta- replied:
> There appears to be a lot of code duplication between this and addUserToGroupDialog.

+1

OrdiNeu commented on modules/principals/src/main/frontend/src/Userboard/Users/changeuserpassworddialogue.jsx:
```diff
@@ -0,0 +1,101 @@
+                                name="newpwdconfirm"
+                                label="Confirm New Password"
+                                onChange={(event) => { this.setState({ newPwdConfirm: event.target.value }); }}
+                            />
```
These three `TextField`s should have `type="password"` passed to them.

OrdiNeu commented on modules/principals/src/main/frontend/src/Userboard/Users/deleteuserdialogue.jsx:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    render() {
+        const { classes } = this.props;
```
No styles are applied, so this is unused.
```suggestion
```

OrdiNeu commented on modules/principals/src/main/frontend/src/Userboard/themeStyle.jsx:
```diff
@@ -0,0 +1,301 @@
+/*!
```
If you rebase off `dev` it should no longer be necessary to copy this file around everywhere. Instead, use the fact that code gets aggregated into aggregated-frontend to pull this copy from elsewhere.

veronikaslc replied:
This sounds like more global issue (like improvement on LFS-248) rather than the scope of this PR. We have this file also use in `data-entry` and `homepage` modules. @marta- Should we make it a separate issue to fix after this one gets merged?

OrdiNeu commented on modules/principals/src/main/frontend/src/Userboard/userboardStyle.jsx:
```diff
@@ -0,0 +1,67 @@
+      verticalAlign: "bottom"
+    },
+    cardTitle: {
+      color: grayColor[2],
```
I think we're currently preferring `theme.palette.grey['200']` over `grayColor`

OrdiNeu commented on modules/principals/src/main/frontend/src/Userboard/userboardStyle.jsx:
```diff
@@ -0,0 +1,67 @@
+      paddingLeft: "120px"
+    },
+    info: {
+      backgroundColor: infoColor[0]
```
`infoColor[0]` should likewise be `theme.palette.info.main`

OrdiNeu reviewed (changes requested):
The code in general needs comments, and the dialogue components appear to reuse lots of code. This may be the case that we need a smaller component that handles the parts of the dialogue similar between all four Create/Delete/AddUser/RemoveUser dialogues. Finally, functional components are preferred over class based ones. 

In all though, the components look good and are pretty functional -- maybe we should consider using MaterialTable elsewhere as well?

veronikaslc committed:
LFS-34: UI for granting user accounts
Syntactic code improvements
Removed unused code

marta- commented on modules/principals/src/main/frontend/src/Userboard/Groups/addusertogroupdialogue.jsx:
```diff
@@ -0,0 +1,189 @@
+    }
+
+    handleAddUsers() {
+        let url = "/system/userManager/group/" + this.props.name + ".update.html";
```
Please pull the service url template in a constant. Duplication makes code difficult to maintain.

marta- commented on modules/principals/src/main/frontend/src/Userboard/Groups/addusertogroupdialogue.jsx:
```diff
@@ -0,0 +1,189 @@
+                let username = data.members[i];
+                username = username.split('/').pop();
+                groupUsers.push(username);
+              }
```
```
var groupUsers = data?.members?.map((n) => n?.split('/').pop());
```

marta- replied:
```
var groupUsers = data?.members?.map((n) => n?.split('/').pop());
```

marta- replied:
```
var groupUsers = data?.members?.map((n) => n?.split('/').pop());
```

marta- commented on modules/principals/src/main/frontend/src/Userboard/Groups/addusertogroupdialogue.jsx:
```diff
@@ -0,0 +1,189 @@
+                groupUsers.push(username);
+              }
+
+              let filtered = [];
```
No need to initialize with an empty array.

``` let filtered = this.state.allUserNames.filter....```

marta- commented on modules/principals/src/main/frontend/src/Userboard/Groups/addusertogroupdialogue.jsx:
```diff
@@ -0,0 +1,189 @@
+                              style={{ boxShadow : 'none' }}
+                              options={
+                                { draggable: false },
+                                { headerStyle: { backgroundColor: '#fafbfc'} },
```
Please use theme colors whenever possible. We should minimize custom styling.

veronikaslc commented:
Most of the codereview comments are addressed. 
ToDo: 
- switch to functional components 
- avoid code duplication between dialogs
- display contents of deleteUserFromGroup dialog instead of group details 
- display user groups table instead of user details

marta- commented on modules/principals/src/main/frontend/src/Userboard/Groups/addusertogroupdialogue.jsx:
```diff
@@ -0,0 +1,181 @@
+                var names = [];
+                for (var username in data) {
+                    names = Object.keys(data).filter((username) => !username.startsWith("sling-")).map((username) => this.addName(username));
+                }
```
The for loop surrounding line 56 looks like a leftover, remove.

```suggestion
                var names = Object.keys(data).filter((username) => !username.startsWith("sling-")).map((username) => this.addName(username));
```

marta- commented on modules/principals/src/main/frontend/src/Userboard/Users/createuserdialogue.jsx:
```diff
@@ -0,0 +1,124 @@
+import React from "react";
+import { Button, Grid, Dialog, DialogTitle, DialogActions, DialogContent, TextField } from "@material-ui/core";
+
+class CreateUserDialogue extends React.Component {
```
You should look into reusing the registration dialog for creating new users.

marta- commented on modules/principals/src/main/frontend/src/Userboard/Users/usersmanager.jsx:
```diff
@@ -0,0 +1,229 @@
+        } else {
+            data.rows[i].initials = data.rows[i].name.charAt(0);
+        }
+      }
```
```suggestion
      data.rows?.forEach((r) => r.initials = (r.firstname?.charAt(0) + r.lastname?.charAt(0)) || r.name?.charAt(0) || '?')
```

veronikaslc committed:
LFS-34: UI for granting user accounts

Co-Authored-By: Marta Girdea <marta.girdea@gmail.com>

veronikaslc committed:
LFS-34: UI for granting user accounts

Co-Authored-By: Marta Girdea <marta.girdea@gmail.com>

veronikaslc committed:
LFS-34: UI for granting user accounts
Display contents of deleteUserFromGroup dialog instead of group details.

veronikaslc committed:
LFS-34: UI for granting user accounts
Display user groups table instead of user details

veronikaslc committed:
LFS-34: UI for granting user accounts
Reuse the registration dialog for creating new users.

veronikaslc committed:
LFS-34: UI for granting user accounts
Do not display any details table if there is no data
Make change password as table action next to delete user
Fix user search in details table for group
Fetch all users and groups on page load
Disable collapse of detals section after any change

Commit:
LFS-34: UI for granting user accounts
Display error on the same create dialog form instead of a new popup

sashaandjic commented:
The handling of the duplicate username works as suggested. Can we do the same for the duplicate group?

Commit:
LFS-34: UI for granting user accounts
Display error for creating new group

sashaandjic commented:
Just one minor complaint. We better display something like "M group already exists." rather than Server error. Everything else is good!
![image](https://user-images.githubusercontent.com/23243244/80619700-0e785400-8a13-11ea-8496-2c83161048d7.png)

veronikaslc commented:
@sashaandjic The error message returned by the backend, this should be changed there. The thing is - already existing could be not the only reason of failure. There can be other problems like authorization, internal errors and such. Since this PR only adressess U I'd leave it for the separate jira issue.

sashaandjic reviewed (approved):
tested

OrdiNeu commented:
Please rebase off `dev`.

OrdiNeu commented on modules/principals/src/main/frontend/src/Userboard/userboardStyle.jsx:
```diff
@@ -0,0 +1,72 @@
+      float: 'right'
+    },
+    dialogTitle: {
+      padding: "16px 0 0 24px"
```
We generally prefer theme.spacing where appropriate.
```suggestion
      padding: theme.spacing(2,0,0,4)
```

OrdiNeu commented on modules/principals/src/main/frontend/src/Userboard/principalsMain.jsx:
```diff
@@ -0,0 +1,101 @@
+import UsersManager from './Users/usersmanager.jsx';
+import GroupsManager from './Groups/groupsmanager.jsx';
+
+class PricipalsContainer extends React.Component {
```
Typo here -- should it be PrincipalsContainer? Will need to edit the export below as well.

OrdiNeu replied:
Also the import in GroupsContainer

OrdiNeu commented on modules/principals/src/main/frontend/src/Userboard/Groups/groupsMain.jsx:
```diff
@@ -0,0 +1,34 @@
+import React from 'react';
+import PricipalsContainer from '../principalsMain.jsx';
+
+class UsersContainer extends React.Component {
```
This file confuses me -- the class is named UsersContainer, the file is named groupsMain, it gets exported and used as groupsmanager in the webpack file. All three names should ideally be the same.

OrdiNeu reviewed (approved):
Works pretty well, and the UI is nice! Just a few minor code comments.

veronikaslc committed:
LFS-34: UI for granting user accounts

Use theme.spacing in styles

Co-authored-by: OrdiNeu <OrdinaryNeumann@gmail.com>

Commit:
LFS-34: UI for granting user accounts
Renamed components and files properly to avoid confusion

Commit:
Merge branch 'LFS-34-rebased' of https://github.com/ccmbioinfo/lfs into LFS-34-rebased

sashaandjic commented:
@veronikaslc  ready for re-test?

veronikaslc commented:
@sashaandjic Yes.

sashaandjic commented:
Should "change password" and "delete user" functions be working in this PR?
- Change password dialog seems to not be working ( I can type anything as a current password, the new passwords are not saved)
- When I tried to login with the wrong password, the screen just went blank.
- I logged as Mary, and deleted myself. The instance went to "coma". 

VM http://172.20.4.116:8080/

veronikaslc commented:
@sashaandjic Changing passwords worked for me as expected. @OrdiNeu Can you please double-ckeck the build for changing user passwords?

sashaandjic commented:
I am sorry for not being more specific. The password change works if you enter proper information.
If you fail to provide existing password or fail to provide the identical entries at least 8 chars long for the new password it leads you to believe that the password has been changed. Then if you try to login with the new password you get the blank screen (this was logged separately).

To sum up:
-  if user enters the wrong existing password, display the error in red fonts
-  if new password isn't at least 8 chars long, display error
-  when new password entries are not identical, the error in red fonts should be displayed

Here is the samples of error msgs taken from admin module ( when creating a new user):
![image](https://user-images.githubusercontent.com/23243244/81530380-bf91bf00-932e-11ea-9014-843997441ec9.png)
![image](https://user-images.githubusercontent.com/23243244/81530849-ae957d80-932f-11ea-9eac-02bc451cafb6.png)

Commit:
LFS-34: UI for granting user accounts
Added same password validation as for login or signup process

veronikaslc commented:
@sashaandjic I have added a password validation. Ready for re-test.

sashaandjic commented:
Looks and works much better. If I enter incorrect old password it will not reset it, however it won't let me know that the change password failed due to the wrong old password.
Right now we only check if the old password is at least 8 chars long. 
The old password option should be presented only when changing your own password. When an administrator resets the user password, they should not need to enter user's old password. To sum up:
- remove old password line for admins
- when changing your own password, check if the old password is valid before allowing any further action
- change the title of the dialog to "Change password for " + username
- once password is successfully changed, close the dialog

Commit:
LFS-34: UI for granting user accounts
Removed old password input from change password dialog

veronikaslc commented:
Addressed all comments, should be perfect now.

Commit:
LFS-34: UI for granting user accounts
Changed wording for change password dialog

sashaandjic reviewed (approved):
tested

Commit:
LFS-34: UI for granting user accounts
Unified dialogs to one style; Code cleanup

sashaandjic reviewed (approved):
Tested. All  is working and looking very good. All dialogs are in line with the LFS standard proposed by Marta.

Commit:
LFS-34: UI for granting user accounts
Code cleanup
